File: app/src/main/java/ch/blitzrechnen/app/generators/Generators.kt
package ch.blitzrechnen.app.generators

import ch.blitzrechnen.app.model.ExerciseType
import ch.blitzrechnen.app.model.Level
import ch.blitzrechnen.app.model.Task
import ch.blitzrechnen.app.model.Visual
import kotlin.random.Random

/**
 * Aufgaben-Generatoren für die 2. Klasse (Zahlenraum bis 100), Lehrplan 21.
 * Jeder Generator liefert eine zufällige [Task] passend zur Schwierigkeitsstufe.
 */
object Generators {

    fun next(type: ExerciseType, level: Level, rnd: Random = Random.Default): Task = when (type) {
        ExerciseType.ANZAHL -> anzahl(level, rnd)
        ExerciseType.TAFEL -> tafel(level, rnd)
        ExerciseType.SCHRITTE -> schritte(level, rnd)
        ExerciseType.ERG_ZEHNER -> ergZehner(level, rnd)
        ExerciseType.ERG_HUNDERT -> ergHundert(level, rnd)
        ExerciseType.DOPPEL -> doppel(level, rnd)
        ExerciseType.PLUS -> plus(level, rnd)
        ExerciseType.MINUS -> minus(level, rnd)
        ExerciseType.HUNDERT_TEILEN -> hundertTeilen(level, rnd)
        ExerciseType.MALREIHEN -> malreihen(level, rnd)
    }

    // 1) Wie viele? — Anzahl am Hunderterfeld erfassen
    private fun anzahl(level: Level, rnd: Random): Task {
        val n = when (level) {
            Level.ANFAENGER -> rnd.nextInt(1, 5) * 10 + rnd.nextInt(0, 10) // klare Zehner+Einer
            Level.KOENNER -> rnd.nextInt(11, 80)
            Level.PROFI -> rnd.nextInt(21, 100)
        }
        return Task(
            type = ExerciseType.ANZAHL,
            question = "Wie viele Punkte?",
            answer = n,
            hint = "Zähle zuerst die vollen Zehnerstreifen, dann die einzelnen Punkte.",
            visual = Visual.HundredField(n)
        )
    }

    // 2) Zahlen finden — fehlende Zahl an der Hundertertafel
    private fun tafel(level: Level, rnd: Random): Task {
        return when (level) {
            Level.ANFAENGER -> {
                val m = rnd.nextInt(2, 98)
                Task(
                    ExerciseType.TAFEL,
                    "Welche Zahl liegt zwischen ${m - 1} und ${m + 1}?",
                    m, "Die Zahl in der Mitte.",
                    Visual.ChartStrip(listOf(m - 1, null, m + 1))
                )
            }
            Level.KOENNER -> {
                val m = rnd.nextInt(1, 90)
                val target = m + 10
                Task(
                    ExerciseType.TAFEL,
                    "Welche Zahl steht ein Feld unter $m?",
                    target, "Ein Feld nach unten heisst +10.",
                    Visual.ChartStrip(listOf(m, null))
                )
            }
            Level.PROFI -> {
                val m = rnd.nextInt(12, 89)
                Task(
                    ExerciseType.TAFEL,
                    "Welche Zahl kommt vor $m?",
                    m - 1, "Eins weniger.",
                    Visual.ChartStrip(listOf(null, m))
                )
            }
        }
    }

    // 3) Zählen in Schritten (2/5/10, vor/rückwärts)
    private fun schritte(level: Level, rnd: Random): Task {
        val step = when (level) {
            Level.ANFAENGER -> listOf(10, 5).random(rnd)
            Level.KOENNER -> listOf(2, 5, 10).random(rnd)
            Level.PROFI -> listOf(2, 5, 10).random(rnd)
        }
        val forward = if (level == Level.ANFAENGER) true else rnd.nextBoolean()
        val start = if (forward) rnd.nextInt(0, 100 - 4 * step)
        else rnd.nextInt(4 * step, 100)
        val seq = (0..2).map { start + (if (forward) it else -it) * step }
        val answer = start + (if (forward) 3 else -3) * step
        val shown = seq.joinToString(", ") { it.toString() } + ", ___"
        val dir = if (forward) "vorwärts" else "rückwärts"
        return Task(
            ExerciseType.SCHRITTE,
            "$shown\n\nWie geht es weiter? ($step-er Schritte $dir)",
            answer,
            "Immer ${if (forward) "+" else "−"}$step rechnen."
        )
    }

    // 4) Ergänzen zum Zehner
    private fun ergZehner(level: Level, rnd: Random): Task {
        val range = when (level) {
            Level.ANFAENGER -> 1..3
            Level.KOENNER -> 1..9
            Level.PROFI -> 1..9
        }
        val tens = rnd.nextInt(1, if (level == Level.ANFAENGER) 5 else 10)
        val ones = range.random(rnd)
        val start = tens * 10 + ones
        val target = (tens + 1) * 10
        return Task(
            ExerciseType.ERG_ZEHNER,
            "$start + ___ = $target",
            target - start,
            "Wie viel fehlt bis zum nächsten Zehner?",
            Visual.HundredField(start)
        )
    }

    // 5) Ergänzen bis 100
    private fun ergHundert(level: Level, rnd: Random): Task {
        val start = when (level) {
            Level.ANFAENGER -> rnd.nextInt(1, 10) * 10 // volle Zehner
            Level.KOENNER -> rnd.nextInt(1, 10) * 5
            Level.PROFI -> rnd.nextInt(1, 100)
        }
        return Task(
            ExerciseType.ERG_HUNDERT,
            "$start + ___ = 100",
            100 - start,
            "Denke: Wie weit ist es von $start bis 100?",
            Visual.HundredField(start)
        )
    }

    // 6) Verdoppeln & Halbieren
    private fun doppel(level: Level, rnd: Random): Task {
        val halve = when (level) {
            Level.ANFAENGER -> false
            else -> rnd.nextBoolean()
        }
        return if (halve) {
            val half = rnd.nextInt(1, if (level == Level.PROFI) 50 else 26)
            val n = half * 2
            Task(ExerciseType.DOPPEL, "Halbiere $n", half, "Die Hälfte von $n.")
        } else {
            val n = when (level) {
                Level.ANFAENGER -> rnd.nextInt(1, 6) * 5   // 5er/10er
                Level.KOENNER -> rnd.nextInt(1, 6) * 5
                Level.PROFI -> rnd.nextInt(10, 50)
            }
            Task(ExerciseType.DOPPEL, "Verdopple $n", n * 2, "$n und noch einmal $n.")
        }
    }

    // 7) Plusaufgaben
    private fun plus(level: Level, rnd: Random): Task {
        return when (level) {
            Level.ANFAENGER -> {
                // ZE + E ohne Übergang
                val tens = rnd.nextInt(1, 9)
                val e1 = rnd.nextInt(0, 5)
                val e2 = rnd.nextInt(0, 5)
                val a = tens * 10 + e1
                Task(ExerciseType.PLUS, "$a + $e2 = ", a + e2, "Zähle die Einer zusammen.")
            }
            Level.KOENNER -> {
                // ZE + ZE ohne Zehnerübergang
                val t1 = rnd.nextInt(1, 5); val e1 = rnd.nextInt(0, 5)
                val t2 = rnd.nextInt(1, 4); val e2 = rnd.nextInt(0, 5 - e1 + 1).coerceAtLeast(0)
                val a = t1 * 10 + e1; val b = t2 * 10 + e2
                Task(ExerciseType.PLUS, "$a + $b = ", a + b, "Zehner und Einer getrennt rechnen.")
            }
            Level.PROFI -> {
                // mit Zehnerübergang
                val a = rnd.nextInt(6, 60)
                val b = rnd.nextInt(6, 100 - a).coerceAtLeast(6)
                Task(ExerciseType.PLUS, "$a + $b = ", a + b, "Erst zum Zehner, dann weiter.")
            }
        }
    }

    // 8) Minusaufgaben
    private fun minus(level: Level, rnd: Random): Task {
        return when (level) {
            Level.ANFAENGER -> {
                val tens = rnd.nextInt(1, 9)
                val e1 = rnd.nextInt(2, 9)
                val a = tens * 10 + e1
                val b = rnd.nextInt(1, e1 + 1)
                Task(ExerciseType.MINUS, "$a − $b = ", a - b, "Nimm nur von den Einern weg.")
            }
            Level.KOENNER -> {
                val t1 = rnd.nextInt(2, 9); val e1 = rnd.nextInt(0, 9)
                val t2 = rnd.nextInt(1, t1); val e2 = rnd.nextInt(0, e1 + 1)
                val a = t1 * 10 + e1; val b = t2 * 10 + e2
                Task(ExerciseType.MINUS, "$a − $b = ", a - b, "Zehner und Einer getrennt.")
            }
            Level.PROFI -> {
                val a = rnd.nextInt(20, 100)
                val b = rnd.nextInt(6, a - 1)
                Task(ExerciseType.MINUS, "$a − $b = ", a - b, "Erst zum Zehner zurück.")
            }
        }
    }

    // 9) 100 teilen
    private fun hundertTeilen(level: Level, rnd: Random): Task {
        val part = when (level) {
            Level.ANFAENGER -> rnd.nextInt(1, 10) * 10
            Level.KOENNER -> rnd.nextInt(1, 20) * 5
            Level.PROFI -> rnd.nextInt(1, 100)
        }
        return Task(
            ExerciseType.HUNDERT_TEILEN,
            "100 = $part + ___",
            100 - part,
            "Die beiden Teile ergeben zusammen 100."
        )
    }

    // 10) Malreihen 2, 5, 10
    private fun malreihen(level: Level, rnd: Random): Task {
        val factor = when (level) {
            Level.ANFAENGER -> listOf(2, 10).random(rnd)
            Level.KOENNER -> listOf(2, 5, 10).random(rnd)
            Level.PROFI -> listOf(2, 5, 10).random(rnd)
        }
        val n = rnd.nextInt(1, 11)
        return Task(
            ExerciseType.MALREIHEN,
            "$factor · $n = ",
            factor * n,
            "Zähle in $factor-er-Schritten $n mal."
        )
    }
}

File: expo/src/generators.ts
import { LevelId, Task } from './models';

// Zufallshilfen (untere Grenze inkl., obere exkl. – wie kotlin Random.nextInt)
function rnd(fromIncl: number, untilExcl: number): number {
  return fromIncl + Math.floor(Math.random() * (untilExcl - fromIncl));
}
function pick<T>(arr: T[]): T {
  return arr[Math.floor(Math.random() * arr.length)];
}

/**
 * Aufgaben-Generatoren für die 2. Klasse (Zahlenraum bis 100), Lehrplan 21.
 * 1:1 portiert aus der nativen Kotlin-App.
 */
export function nextTask(typeId: string, level: LevelId): Task {
  switch (typeId) {
    case 'anzahl': return anzahl(level);
    case 'tafel': return tafel(level);
    case 'schritte': return schritte(level);
    case 'erg_zehner': return ergZehner(level);
    case 'erg_hundert': return ergHundert(level);
    case 'doppel': return doppel(level);
    case 'plus': return plus(level);
    case 'minus': return minus(level);
    case 'hundert_teilen': return hundertTeilen(level);
    case 'malreihen': return malreihen(level);
    default: return plus(level);
  }
}

function anzahl(level: LevelId): Task {
  let n: number;
  if (level === 0) n = rnd(1, 5) * 10 + rnd(0, 10);
  else if (level === 1) n = rnd(11, 80);
  else n = rnd(21, 100);
  return {
    typeId: 'anzahl',
    question: 'Wie viele Punkte?',
    answer: n,
    hint: 'Zähle zuerst die vollen Zehnerstreifen, dann die einzelnen Punkte.',
    visual: { kind: 'hundredField', count: n },
  };
}

function tafel(level: LevelId): Task {
  if (level === 0) {
    const m = rnd(2, 98);
    return { typeId: 'tafel', question: `Welche Zahl liegt zwischen ${m - 1} und ${m + 1}?`, answer: m, hint: 'Die Zahl in der Mitte.', visual: { kind: 'chartStrip', values: [m - 1, null, m + 1] } };
  } else if (level === 1) {
    const m = rnd(1, 90);
    return { typeId: 'tafel', question: `Welche Zahl steht ein Feld unter ${m}?`, answer: m + 10, hint: 'Ein Feld nach unten heisst +10.', visual: { kind: 'chartStrip', values: [m, null] } };
  } else {
    const m = rnd(12, 89);
    return { typeId: 'tafel', question: `Welche Zahl kommt vor ${m}?`, answer: m - 1, hint: 'Eins weniger.', visual: { kind: 'chartStrip', values: [null, m] } };
  }
}

function schritte(level: LevelId): Task {
  const step = level === 0 ? pick([10, 5]) : pick([2, 5, 10]);
  const forward = level === 0 ? true : Math.random() < 0.5;
  const start = forward ? rnd(0, 100 - 4 * step) : rnd(4 * step, 100);
  const seq = [0, 1, 2].map((i) => start + (forward ? i : -i) * step);
  const answer = start + (forward ? 3 : -3) * step;
  const shown = seq.join(', ') + ', ___';
  const dir = forward ? 'vorwärts' : 'rückwärts';
  return { typeId: 'schritte', question: `${shown}\n\nWie geht es weiter? (${step}er-Schritte ${dir})`, answer, hint: `Immer ${forward ? '+' : '−'}${step} rechnen.` };
}

function ergZehner(level: LevelId): Task {
  const tens = rnd(1, level === 0 ? 5 : 10);
  const ones = level === 0 ? rnd(1, 4) : rnd(1, 10);
  const start = tens * 10 + ones;
  const target = (tens + 1) * 10;
  return { typeId: 'erg_zehner', question: `${start} + ___ = ${target}`, answer: target - start, hint: 'Wie viel fehlt bis zum nächsten Zehner?', visual: { kind: 'hundredField', count: start } };
}

function ergHundert(level: LevelId): Task {
  let start: number;
  if (level === 0) start = rnd(1, 10) * 10;
  else if (level === 1) start = rnd(1, 10) * 5;
  else start = rnd(1, 100);
  return { typeId: 'erg_hundert', question: `${start} + ___ = 100`, answer: 100 - start, hint: `Denke: Wie weit ist es von ${start} bis 100?`, visual: { kind: 'hundredField', count: start } };
}

function doppel(level: LevelId): Task {
  const halve = level === 0 ? false : Math.random() < 0.5;
  if (halve) {
    const half = rnd(1, level === 2 ? 50 : 26);
    const n = half * 2;
    return { typeId: 'doppel', question: `Halbiere ${n}`, answer: half, hint: `Die Hälfte von ${n}.` };
  }
  const n = level === 2 ? rnd(10, 50) : rnd(1, 6) * 5;
  return { typeId: 'doppel', question: `Verdopple ${n}`, answer: n * 2, hint: `${n} und noch einmal ${n}.` };
}

function plus(level: LevelId): Task {
  if (level === 0) {
    const tens = rnd(1, 9), e1 = rnd(0, 5), e2 = rnd(0, 5);
    const a = tens * 10 + e1;
    return { typeId: 'plus', question: `${a} + ${e2} = `, answer: a + e2, hint: 'Zähle die Einer zusammen.' };
  } else if (level === 1) {
    const t1 = rnd(1, 5), e1 = rnd(0, 5);
    const t2 = rnd(1, 4), e2 = rnd(0, 5 - e1 + 1);
    const a = t1 * 10 + e1, b = t2 * 10 + e2;
    return { typeId: 'plus', question: `${a} + ${b} = `, answer: a + b, hint: 'Zehner und Einer getrennt rechnen.' };
  } else {
    const a = rnd(6, 60);
    const b = Math.max(6, rnd(6, 100 - a));
    return { typeId: 'plus', question: `${a} + ${b} = `, answer: a + b, hint: 'Erst zum Zehner, dann weiter.' };
  }
}

function minus(level: LevelId): Task {
  if (level === 0) {
    const tens = rnd(1, 9), e1 = rnd(2, 9);
    const a = tens * 10 + e1;
    const b = rnd(1, e1 + 1);
    return { typeId: 'minus', question: `${a} − ${b} = `, answer: a - b, hint: 'Nimm nur von den Einern weg.' };
  } else if (level === 1) {
    const t1 = rnd(2, 9), e1 = rnd(0, 9);
    const t2 = rnd(1, t1), e2 = rnd(0, e1 + 1);
    const a = t1 * 10 + e1, b = t2 * 10 + e2;
    return { typeId: 'minus', question: `${a} − ${b} = `, answer: a - b, hint: 'Zehner und Einer getrennt.' };
  } else {
    const a = rnd(20, 100);
    const b = rnd(6, a - 1);
    return { typeId: 'minus', question: `${a} − ${b} = `, answer: a - b, hint: 'Erst zum Zehner zurück.' };
  }
}

function hundertTeilen(level: LevelId): Task {
  let part: number;
  if (level === 0) part = rnd(1, 10) * 10;
  else if (level === 1) part = rnd(1, 20) * 5;
  else part = rnd(1, 100);
  return { typeId: 'hundert_teilen', question: `100 = ${part} + ___`, answer: 100 - part, hint: 'Die beiden Teile ergeben zusammen 100.' };
}

function malreihen(level: LevelId): Task {
  const factor = level === 0 ? pick([2, 10]) : pick([2, 5, 10]);
  const n = rnd(1, 11);
  return { typeId: 'malreihen', question: `${factor} · ${n} = `, answer: factor * n, hint: `Zähle in ${factor}er-Schritten ${n} mal.` };
}
